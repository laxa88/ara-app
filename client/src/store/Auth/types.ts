export const enum ActionTypes {
  LOGIN = 'AUTH/LOGIN',
  LOGIN_SUCCESS = 'AUTH/LOGIN_SUCCESS',
  LOGIN_FAIL = 'AUTH/LOGIN_FAIL',
  LOGOUT = 'AUTH/LOGOUT',
}

export interface IAction {
  type: string;
  payload?: {
    email?: string;
    password?: string;
    message?: string;
    token?: string;
  };
  error?: boolean;
  meta?: any;
}

export interface IState {
  isLoggedIn: boolean;
  isLoading: boolean;
  errorMessage: string;
}

export interface IProps {
  isLoggedIn: boolean;
  isLoading: boolean;
  errorMessage: string;
}

export interface IDispatch {
  login?: (email: string, password: string) => void;
  logout?: () => void;
}
