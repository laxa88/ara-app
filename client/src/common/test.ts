import { configure, mount, render, shallow } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import reduxMockStore from 'redux-mock-store';

configure({ adapter: new Adapter() });

// Reserved for later on, if we use thunk or saga or anything else
const middleware: any = [];

function mockStore(state: any) {
  return reduxMockStore(middleware)(state);
}

export { shallow, mount, render, mockStore };
