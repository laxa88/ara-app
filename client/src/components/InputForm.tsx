import * as React from 'react';

import Form from './Form';
import Input, { IProps as IInput } from './Input';

export interface IProps extends IInput {
  label: string;
}

class InputForm extends React.Component<IProps, {}> {
  constructor(props: IProps) {
    super(props);
  }

  public render() {
    const { label, value, onChange, ...others } = this.props;

    return (
      <Form label={label}>
        <Input value={value} onChange={onChange} {...others} />
      </Form>
    );
  }
}

export default InputForm;
