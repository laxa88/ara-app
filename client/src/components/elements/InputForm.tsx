import * as React from 'react';
import styled from '../css/styled';

import Input, { IInput } from './Input';

const Label = styled.label`
  color: red;
`;

export interface IProps extends IInput {
  label: string;
}

class InputForm extends React.Component<IProps, {}> {
  constructor(props: IProps) {
    super(props);
  }

  public render() {
    const { label, value, onChange, ...others } = this.props;

    return (
      <form>
        <Label>{label}</Label>
        <Input value={value} onChange={onChange} {...others} />
      </form>
    );
  }
}

export default InputForm;
