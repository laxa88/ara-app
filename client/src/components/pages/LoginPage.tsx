import * as React from 'react';

import Input from '../elements/Input';

class LoginPage extends React.Component<{}, {}> {
  constructor(props: {}) {
    super(props);
  }

  public render() {
    return (
      <div>
        <Input type="email" placeholder="Input email here" />
        <Input placeholder="Input password here" />
      </div>
    );
  }
}

export default LoginPage;
