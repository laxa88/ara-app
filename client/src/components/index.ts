import Input from './Input';
import InputForm from './InputForm';

export { Input, InputForm };
