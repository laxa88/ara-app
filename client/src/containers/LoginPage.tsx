import * as React from 'react';
import { connect } from 'react-redux';
import * as Redux from 'redux';

import Button from '../components/Button';
import InputForm from '../components/InputForm';
import * as loginPageActions from '../store/LoginPage/actions';
import { IDispatch, IProps, IState } from '../store/LoginPage/types';

type Props = IProps & IDispatch;

export class LoginPage extends React.Component<Props, {}> {
  constructor(props: Props) {
    super(props);
  }

  public render() {
    return (
      <div>
        <InputForm
          label="Email"
          type="email"
          value={this.props.email}
          placeholder="Input email here"
          onChange={this.onChangeEmail}
        />
        <InputForm
          label="Password"
          value={this.props.password}
          type="password"
          placeholder="Input password here"
          onChange={this.onChangePassword}
        />
        <Button onClick={this.onClickLogin}>Login</Button>
      </div>
    );
  }

  private onClickLogin = () => {
    this.props.login();
  }

  private onChangeEmail = (value: string) => {
    this.props.setEmail(value);
  }

  private onChangePassword = (value: string) => {
    this.props.setPassword(value);
  }
}

const mapStateToProps = (state: IState): IProps => ({
  email: state.email,
  password: state.password,
});

const mapDispatchToProps = (dispatch: Redux.Dispatch): IDispatch => ({
  login: () => dispatch(loginPageActions.login()),
  setEmail: (val: string) => dispatch(loginPageActions.setEmail(val)),
  setPassword: (val: string) => dispatch(loginPageActions.setPassword(val)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(LoginPage);
