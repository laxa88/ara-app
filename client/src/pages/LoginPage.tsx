import * as React from 'react';
import { connect } from 'react-redux';

import InputForm from '../components/InputForm';
import * as loginPageActions from '../store/LoginPage/actions';
import { IDispatch, IState } from '../store/LoginPage/types';

interface IDispatchMap {
  setEmail: (email: string) => void;
  setPassword: (password: string) => void;
}

interface IProps extends IDispatchMap {
  email: string;
  password: string;
}

class LoginPage extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);

    this.onChangeEmail = this.onChangeEmail.bind(this);
    this.onChangePassword = this.onChangePassword.bind(this);
  }

  public render() {
    return (
      <div>
        <InputForm
          label="Email"
          type="email"
          value={this.props.email}
          placeholder="Input email here"
          onChange={this.onChangeEmail}
        />
        <InputForm
          label="Password"
          value={this.props.password}
          type="password"
          placeholder="Input password here"
          onChange={this.onChangePassword}
        />
      </div>
    );
  }

  private onChangeEmail(value: string) {
    this.props.setEmail(value);
  }

  private onChangePassword(value: string) {
    this.props.setPassword(value);
  }
}

const mapStateToProps = (state: IState): IProps => ({
  email: state.email,
  password: state.password,
  setEmail: () => undefined,
  setPassword: () => undefined,
});

const mapDispatchToProps = (dispatch: IDispatch): IDispatchMap => ({
  setEmail: (val: string) => dispatch(loginPageActions.setEmail(val)),
  setPassword: (val: string) => dispatch(loginPageActions.setPassword(val)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(LoginPage);
