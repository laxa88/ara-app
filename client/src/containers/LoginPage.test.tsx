import * as React from 'react';
import { Provider } from 'react-redux';

import { mockStore, mount, shallow } from '../common/test';
import { IProps as IInputForm } from '../components/InputForm';
import { ActionTypes } from '../store/LoginPage/types';
import LoginPage, { LoginPage as Base } from './LoginPage';

describe('<LoginPage />', () => {
  it('should render', () => {
    const mockProps = {
      email: 'dummy email',
      login: jest.fn(),
      password: 'dummy password',
      setEmail: jest.fn(),
      setPassword: jest.fn(),
    };

    const wrapper = shallow(<Base {...mockProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should update email and password', () => {
    // Arrange

    const mockEmail1 = 'dummy email';
    const mockPassword1 = 'dummy password';

    const mockEmail2 = 'new dummy email';
    const mockPassword2 = 'new dummy password';

    const initialState = {
      email: mockEmail1,
      password: mockPassword1,
    };

    const expected = [
      { type: ActionTypes.SET_EMAIL, payload: { email: mockEmail2 } },
      { type: ActionTypes.SET_PASSWORD, payload: { password: mockPassword2 } },
    ];

    const store = mockStore(initialState);

    const wrapper = mount(
      <Provider store={store}>
        <LoginPage />
      </Provider>,
    );

    const inputForms = wrapper.children().find('InputForm');
    type InputFormProps = React.HTMLAttributes<HTMLInputElement> & IInputForm;

    // Act

    (inputForms.at(0).props() as InputFormProps).onChange(mockEmail2, '');
    (inputForms.at(1).props() as InputFormProps).onChange(mockPassword2, '');

    // Assert

    expect(store.getActions()).toEqual(expected);
  });
});
