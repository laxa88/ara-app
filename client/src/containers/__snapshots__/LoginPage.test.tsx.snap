// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage /> should render 1`] = `
<div>
  <InputForm
    label="Email"
    onChange={[Function]}
    placeholder="Input email here"
    type="email"
    value="dummy email"
  />
  <InputForm
    label="Password"
    onChange={[Function]}
    placeholder="Input password here"
    type="password"
    value="dummy password"
  />
  <Button
    onClick={[Function]}
  >
    Login
  </Button>
</div>
`;
